templates/fargate-service.cfn.yml

```diff
@@ -0,0 +1,682 @@
+---
+AWSTemplateFormatVersion: 2010-09-09
+
+# A CloudFormation template to deploy an additional service to Fargate. This requires an existing
+# cluster deployed by fargate.cfn.yml.
+
+Description: Fargate Service
+
+
+Parameters:
+
+  NetworkStackName:
+    Type: String
+    Description: Name of an active Startup Kit CloudFormation stack that contains networking resources
+    MinLength: 1
+    MaxLength: 255
+    AllowedPattern: "^[a-zA-Z][-a-zA-Z0-9]*$"
+
+  EnvironmentName:
+    Type: String
+    Description: Environment name - dev or prod
+    Default: dev
+    AllowedValues:
+      - dev
+      - prod
+    ConstraintDescription: Specify either dev or prod
+
+  DatabaseStackName:
+    Type: String
+    Description: Name of an optional active Startup Kit CloudFormation stack that contains database resources
+    Default: ""
+
+  FargateStackName:
+    Type: String
+    Description: Name of an active Startup Kit CloudFormation stack that contains Fargate resources
+    MinLength: 1
+    MaxLength: 255
+    AllowedPattern: "^[a-zA-Z][-a-zA-Z0-9]*$"
+
+  RegisterServiceWithAlb:
+    Default: true
+    Type: String
+    Description: Set to false to disable registering with the ALB (e.g., backend services without a web interface)
+    ConstraintDescription: Only true or false are allowed
+    AllowedValues:
+      - true
+      - false
+
+  AppProtocol:
+    Type: String
+    Description: The application server protocol
+    Default: HTTP
+    AllowedValues:
+      - HTTP
+      - HTTPS
+    ConstraintDescription: Specify either HTTTP or HTTPS
+
+  ServiceURLPath:
+    Type: String
+    Description: The URL path for the service (e.g., /test)
+    Default: /test
+    MinLength: 1
+    MaxLength: 255
+    ConstraintDescription: Value must be between 1 and 255 characters
+
+  ServiceLBListenerPriority:
+    Type: Number
+    Description: The service load balancer listener priority - must be unique for each service if ALB is enabled
+    Default: 2
+    MinValue: 2
+    MaxValue: 50000
+    ConstraintDescription: Number must be between 1 and 50,000
+
+  HealthCheckPath:
+    Type: String
+    Description: The path for the Application Load Balancer health check
+    Default: /
+    MinLength: 1
+    MaxLength: 255
+    ConstraintDescription: Value must be between 1 and 255 characters
+
+  GitHubSourceRepo:
+    Type: String
+    Description: GitHub source repository - must contain a Dockerfile in the base
+
+  GitHubBranch:
+    Type: String
+    Default: master
+    Description: GitHub repository branch to trigger builds
+
+  GitHubToken:
+    Type: String
+    NoEcho: true
+    Description: "GitHub API token - see: https://github.com/blog/1509-personal-api-tokens"
+
+  GitHubUser:
+    Type: String
+    Description: GitHub username
+
+  CodeBuildDockerImage:
+    Type: String
+    Default: aws/codebuild/docker:17.09.0
+
+  SeedDockerImage:
+    Type: String
+    Default: registry.hub.docker.com/library/nginx:1.13
```

Comment:
This still sucks but I don't have a better suggestion. My main problem here is how do we update this stack post creation without taking down our application? I'm tracking this in the refarch repo as an open issue. Increasingly I'm uncomfortable with CloudFormation stacks that immediately drift and are unable to be updated safely.

Reply:
We can discuss tomorrow, but I do not see a reason why you cannot change this parameter in an update to use your current image. It is a bit clunky because you have to do the bookkeeping, but given the circular dependency in TaskDefinition, it is required.

Reply:
That is a massive sharp edge to leave on these projects. A gun pointed at the foot of the beginners using these projects to get going. I think we can do better.

Reply:
I have not heard a single customer take issue with this. It would be ideal if we didn't have an image dependency in the resource creation, but that is how it works. We are scheduled to discuss this on the phone today and can discuss other options if there are any that don't require custom resources.

Reply:
I have – see https://github.com/awslabs/ecs-refarch-continuous-deployment/issues/33 where I'm tracking it. That said, I wouldn't block yourself on this point though I would track it somewhere. The two options I'm considering are a custom resource (which I don't think will work) or reverting back to CloudFormation to manage infrastructure configuration .